src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -4,67 +4,196 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using Microsoft.Extensions.ObjectPool;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Razor.PooledObjects;
 
 /// <summary>
 ///  Wraps a pooled <see cref="ImmutableArray{T}.Builder"/> but doesn't allocate it until
 ///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
 ///  to the pool.
+///
+///  There is significant effort to avoid retrieving the <see cref="ImmutableArray{T}.Builder"/>.
+///  For very small arrays of length 4 or less, the elements will be stored on the stack. If the array
+///  grows larger than 4 elements, a builder will be employed.
 /// </summary>
-internal ref partial struct PooledArrayBuilder<T>
+internal partial struct PooledArrayBuilder<T> : IDisposable
```

Comment:
Because the Dispose pattern only works for ref structs. If a non-ref struct doesn't implement `IDisposeable`, it can't be used in a using statement.

Reply:
I was running into trouble getting the enumerator to work, since we do that on a copy of the struct. In general, keeping it a ref struct was more trouble than it was worth, since that meant it couldn't be used in async methods.

Reply:
`ref structs` have a lot of limitations that make the implementation inconvenient. For example, If `PooledArrayBuilder<T>` were a ref struct, it could not have a static `Empty` member. It also becomes difficult to implement the enumerator which stores a copy of the `PooledArrayBuilder<T>` as an instance field, so that too must become a ref struct, but that causes other issues with ref safety rules. In essence, I found that it was too awkward to keep this a ref struct and it was already painful that this type couldn't be used in an async method. It does mean that we have to be careful not to capture it in lambdas or local methods because it will work incorrectly, but that's not the normal usage scenario.

Reply:
💡 I would suggest making this `[NonCopyable]` and ensuring the corresponding analyzers are enabled.